state/protocol/protocol_state/kvstore/interfaces.go

```diff
@@ -0,0 +1,48 @@
+package kvstore
+
+// This file contains versioned read and read-write interfaces to the Protocol State
+// key-value store and are used by the Protocol State Machine.
+//
+// When a key is added or removed, this requires a new protocol state version:
+//  - Create a new versioned model in models.go (eg. modelv3 if latest model is modelv2)
+//  - Update the Reader and API interfaces to include any new keys
+
+// Reader is the latest read-only interface to the Protocol State key-value store
+// at a particular block.
+// It is backward-compatible with all versioned model types defined in the models.go
+// for this software version.
+type Reader interface {
```

Comment:
I have a slight preference building `Reader` using dedicated interfaces for each version, it might be useful in some cases where we have components fixed to some version + it's clear what fields were added in which version. Not a strong preference though.

Reply:
I actually started out with public versioned interfaces for the reader, but removed them in the end because they didn't seem strictly necessary. I see the benefits of some kind of compartmentalization by version though. Alex and I were talking about a similar topic earlier today (re https://github.com/onflow/flow-go/pull/5371#discussion_r1487237515), that we partition the interface by "feature" or by state-machine. 

Minimally we should create some best practices and an example for how to organize state in the store, for a given feature. Either way though I think it is best to make that decision as part of dog-fooding the kv store the first time (most likely for Consensus Node HCU).